src/main/java/com/cloudbees/jenkins/plugins/bitbucket/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -139,17 +140,32 @@ public SCMFileSystem build(@NonNull SCMSource source, @NonNull SCMHead head, @Ch
             BitbucketAuthenticator authenticator = AuthenticationTokens.convert(BitbucketAuthenticator.authenticationContext(serverUrl), credentials);
 
             BitbucketApi apiClient = BitbucketApiFactory.newInstance(serverUrl, authenticator, owner, repository);
-            String ref;
+            String ref = null;
+
             if (head instanceof BranchSCMHead) {
                 ref = head.getName();
             } else if (head instanceof PullRequestSCMHead) {
+                // working on a pull request - can be either "HEAD" or "MERGE"
                 PullRequestSCMHead pr = (PullRequestSCMHead) head;
-                if (!(pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) && pr.getRepository() != null) {
-                    return new BitbucketSCMFileSystem(apiClient, pr.getOriginName(), rev);
+                if (pr.getRepository() == null) { // check access to repository (might be forked)
+                    return null;
+                }
+
+                if (apiClient instanceof BitbucketCloudApiClient) {
+                    // Bitbucket cloud does not support refs for pull requests
+                    // TODO waiting for cloud support: https://bitbucket.org/site/master/issues/5814/refify-pull-requests-by-making-them-a-ref
+                    if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.MERGE) {
+                        return null;
+                    } else if (pr.getCheckoutStrategy() == ChangeRequestCheckoutStrategy.HEAD) {
+                        ref = pr.getOriginName();
```

Comment:
Testing out whether this is even possible

Reply:
Yup only possible when the PR is coming from the same repository.
Atlassian is the one to blame for having an issue open for 7 years